java-checks/src/main/java/org/sonar/java/checks/HardcodedURICheck.java

```diff
@@ -117,26 +187,34 @@ private static boolean isFileNameVariable(@Nullable IdentifierTree variable) {
     return variable != null && VARIABLE_NAME_PATTERN.matcher(variable.name()).find();
   }
 
-  private void checkExpression(@Nullable ExpressionTree expr) {
-    if (expr != null) {
-      if (isHardcodedURI(expr)) {
-        reportHardcodedURI(expr);
-      } else {
-        reportStringConcatenationWithPathDelimiter(expr);
-      }
+  private void checkExpression(ExpressionTree expr) {
+    if (isHardcodedURI(expr)) {
+      reportHardcodedURI(expr);
+    } else {
+      reportStringConcatenationWithPathDelimiter(expr);
     }
+
   }
 
   private static boolean isHardcodedURI(ExpressionTree expr) {
```

Comment:
I simplified things a bit with `@Nullable` :
```
  private static boolean isHardcodedURI(ExpressionTree expr) {
    String literal = stringLiteral(expr);
    return literal != null
      && !literal.contains("*")
      && !literal.contains("$")
      && URI_PATTERN.matcher(literal).find();
  }

  @Nullable
  private static String stringLiteral(ExpressionTree expr) {
//    return Optional.of(expr)
//      .map(ExpressionUtils::skipParentheses)
//      .filter(LiteralTree.class::isInstance)
//      .map(e -> (LiteralTree) e)
//      .map(LiteralTree::value)
//      .map(LiteralUtils::trimQuotes);
    ExpressionTree unquoted = ExpressionUtils.skipParentheses(expr);
    if (unquoted instanceof LiteralTree literalTree) {
      return LiteralUtils.trimQuotes(literalTree.value());
    }
    return null;
  }
```

If you prefer Optional then there's a streams version in the comment.

Reply:
indeed code using null instead of optional is shorter and easier to read and the nullable annotations allows type checker to raise an error